### ISSUE TITLE
[Vote] Softmax and Loss Convention

### ISSUE THREAD

**Issue by tqchen** [dmlc/mxnet#434]
According to discussions in #426 
We agreed in all case, the multi-output and single output softmax can be combined into one-class, maybe overloaded by shape of label.

- ***Option 1***
  - ```SoftmaxOutput``` for output without gradient attached.
  - After attach a loss, XXXOutput will be able to backprop gradient of the loss, while forward behavior remains unchanged.
  - ```Softmax``` remains the same, with loss already attached.
  - May need to introduce attach_loss and maybe special loss class.
- ***Option 2***
   - ```SoftmaxOutput``` for output, with specific backward behavior of cross-entropy-loss
       - As an alternative task based naming instead of math based: ```MulticlassProbOutput``` to be clear to user.
   - ```Softmax``` behaves normally(only take input), and can prop gradient back from any output source(being able to compose as internal node)
   - ```CrossEntropyLoss``` can be composed with anything, including ```Softmax```, to get loss in forward and gradient in backward.

Please edit this post to add more options .

**Comment by tqchen** [comment]
I vote for 2

**Comment by piiswrong** [comment]
I vote for 2 with the following name change:
SoftmaxOutput -> SoftmaxLoss
Softmax -> activationop with option "softmax" since its really just an activation

**Comment by antinucleon** [comment]
+1 for 2

**Comment by tqchen** [comment]
@pluskid @mli

**Comment by mli** [comment]
+1 for 2

**Comment by pluskid** [comment]
Sorry currently traveling. I think 1 is a more unified behavior but requires more changes to the internals (eg attach-loss etc). I'm generally ok with either design.

**Comment by tqchen** [comment]
#444  #450 Imposes the changes for python and R

**Comment by tqchen** [comment]
@hjk41  can you make a windows build for the most recent version?

**Comment by tqchen** [comment]
all sides updated
